lampe/inference.py

```diff
@@ -524,14 +540,183 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
+
+    .. math:: \arg\min_\phi \mathbb{E}_{p(\theta, x) p(t) p(\theta_t | \theta)}
+        \Big[ \big\|
+            s_\phi(\theta_t, x, t) -
+            (1 - \alpha(t)) \nabla_{\! \theta_t} \log p(\theta_t | \theta)
+        \big\|_2^2 \Big] .
+
+    For log-density computation and sampling, we solve the probability flow ODE
+
+    .. math:: \mathrm{d} \theta = \left[ -\frac{1}{2} \beta(t) \theta -
+        \frac{1}{2} \beta(t) (1 + \alpha(t)) s_\phi(\theta, x, t) \right] \, \mathrm{d} t
+
+    whose trajectories share the same marginal probability densities :math:`p(\theta_t | x)`
+    as the stochastic diffusion process.
+
+    References:
+        | Score-Based Generative Modeling through Stochastic Differential Equations (Song et al., 2021)
+        | https://arxiv.org/abs/2011.13456
+
+    Arguments:
+        theta_dim: The dimensionality :math:`D` of the parameter space.
+        x_dim: The dimensionality :math:`L` of the observation space.
+        moments: The parameters moments :math:`\mu` and :math:`\sigma`. If provided,
+            the moments are used to standardize the parameters.
+        build: The network constructor (e.g. :class:`lampe.nn.ResMLP`).
+        kwargs: Keyword arguments passed to the constructor.
+    """
+
+    def __init__(
+        self,
+        theta_dim: int,
+        x_dim: int,
+        moments: Tuple[Tensor, Tensor] = None,
+        build: Callable[[int, int], nn.Module] = MLP,
+        **kwargs,
+    ):
+        super().__init__()
+
+        if moments is None:
+            self.standardize = nn.Identity()
+        else:
+            mu, sigma = moments
+            self.standardize = Affine(-mu / sigma, 1 / sigma, trainable=False)
+
+        self.net = build(theta_dim + x_dim + 2, theta_dim, **kwargs)
+
+        self.register_buffer('zeros', torch.zeros(theta_dim))
+        self.register_buffer('ones', torch.ones(theta_dim))
+
+    def forward(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        r"""
+        Arguments:
+            theta: The parameters :math:`\theta`, with shape :math:`(*, D)`.
+            x: The observation :math:`x`, with shape :math:`(*, L)`.
+            t: The time :math:`t`, with shape :math:`(*,).`
+
+        Returns:
+            The score :math:`s_\phi(\theta, x, t)`, with shape :math:`(*,)`.
+        """
+
+        theta = self.standardize(theta)
+        t = torch.stack((
+            torch.cos(math.pi * t),
+            torch.sin(math.pi * t),
+        ), dim=-1)
+        theta, x, t = broadcast(theta, x, t, ignore=1)
+
+        return self.net(torch.cat((theta, x, t), dim=-1))
+
+    def alpha(self, t: Tensor) -> Tensor:
+        return torch.exp(-8.0 * t**2)
+
+    def beta(self, t: Tensor) -> Tensor:
+        return 16.0 * t
+
+    def ode(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        drift = -self.beta(t) / 2 * theta
+        diffusion = -self.beta(t) / 2 * (1 + self.alpha(t)) * self.forward(theta, x, t)
```

Comment:
Shouldn't we have an additional `(1 - self.alpha(t))` factor here?

Reply:
I don't think so. I used a rescaled version (of my own) of the score-matching objective such that

$$s_\phi(\theta, x, t) \approx (1 - \alpha(t)) \nabla_\theta ~ p_t(\theta | x)$$

Thanks to that the variance of the estimator is uniform with respect to the time, which makes the training and ODE more stable (at least in my experiments).

Reply:
Gotcha, I missed the `(rescaled)` in the docstring.